### PR TITLE
add from_model + test

### PR DESCRIPTION
**Pull request by dakl**
Here's a more proper PR.

Please let me know what you think, if this is a feasible way forward or not.

### CODE REVIEW

**Comment by dakl** [comment]
The build seems to have failed due to OOM. I'm not sure that's related to my changes?

**Comment by dakl** [comment]
I fixed the flake8 stuff and the memory thing went away. 🎉

**Comment by maciejkula** [comment]
(OOM sadly happens every now and then. I can always re-run it.)

I like the idea, and thank you for the contribution! I am keen to have something like this in the package, but I would suggest a slightly different API. Maybe something like this:

```
original = (...some model...)
new = original.resize(new_num_user_features, new_num_item_features)
````

As in your suggestion, `new` would be a new, wholly independent object. I am also (rather strongly) opposed to having any type of feature names in the model; I'd rather leave this sort of plumbing to the user. Is there a use case that your suggestion supports where the names are important?

Finally,I think it would be good to carry over the rest of the state of the model, like the accumulated gradients, and all the hyperparameters. The `get_params` and `set_params` method will be useful for the latter.

To sum up: I look forward to having this feature, but I am keen to get it at least somewhat right, so it may take us a couple of iterations to get there.

**Comment by dakl** [comment]
Thanks for the feedback!

I definitely agree with carrying more of the state, I just didn't want to invest too much time in the implementation before getting some feedback. 

I think feature names would be useful in the model. An example could be item features which could be words in the titles of products on an ecommerce site, for example 'cable', 'case' and 'iphone'. Let's say we want to add the feature 'charger' because the item ´iphone charger´ is new in the inventory. This could then be added as the last element in the list, but in preprocessing would be far more likely to be put in the list of features in alphabetical order, resulting in the feature list 'cable', 'case', 'charger' and 'iphone'. Having the words in `item_feature_names` greatly simplifies the creation of a new model for the user, who won't have to care at all in which order the features sit in the model, that's all handled internally in the model. 

I understand that it's of course possible to let the user write the logic for it, but I think it's much cleaner to write in once (in the lightfm implementation) and add proper tests and have users use it and feel confident that it's correct. This kind of logic/plumbing is pretty error prone in my experience, and can require a fair bit work to get right. In the model however, I think it's much eaiser since everything is known in one place in the model object.

Those would be my two cents for having the feature names in the model. 

Regarding the API, I'm fine with `original.resize(...)`. `original.update_features(...)` could be another suggestion. I have no real opinion here, the static method was just the first thing that came to mind :)

Great work with LightFM btw, it's a really useful system! 

cheers

**Comment by maciejkula** [comment]
I see the merit in making things easier for the user. I see a fair number of questions about how to construct features, and having to map IDs people use in their systems to contiguous number ranges also generates a fair amount of confusion. I think it may be time to tackle this more head-on; I'm going to write a quick prototype, share it soon, and maybe then we can collaborate on that?

**Comment by maciejkula** [comment]
This is not to say I don't like this approach, we will definitely get this done too :) But there is a meatier bit about data processing that you correctly suggest should be handled, and I think once we get that done this will become a fair bit easier to do in a nice way.

**Comment by dakl** [comment]
Collaborating on that sounds great. I’ll pause the development of this pull request until your prototype is further along.

**Comment by dakl** [comment]
@maciejkula It strikes me that the `.resize(...)` API might not be intuitive for the user. Having the method be an instance method to me indicates that it's supposed to modify (resize) the instance in question, but in our case it will return a new object, factory method style. I think this was why I chose the `@staticmethod` in my original implementation.

**Comment by dakl** [comment]
As you can see I couldn't stop myself from doing the rest of the implementation. 😜

**Comment by maciejkula** [comment]
I had a go at adding some functionality to making managin user/item ids and feature names easier: https://github.com/lyst/lightfm/pull/281.

Please have a look: I think with that we can offload the responsibility for maintaining consistent features to the `Dataset` class, and change the `resize` function to not accept feature names, but rather the desired dimensions of the new model (as returned by the `model_dimensions` method).

This doesn't allow dropping features (only adding new ones works) --- but I think I'm OK with that.

**Comment by maciejkula** [comment]
(Also, I think `resize` returning a new object is a good idea. Thanks for implementing that.)

**Comment by maciejkula** [comment]
Hey @dakl do you want to collaborate on the design for the data builders?

**Comment by dakl** [comment]
Sorry for the long absence of response. This is no longer relevant, right?
